### PR TITLE
[AutoDiff] Unwrap the top level of linear map tuple when it is possible

### PR DESCRIPTION
**Pull request by asl**
This essentially passes the members of a linear map tuple as individual arguments. It yields few nice simplifications:
  - No linear map tuples at all for getters / setters
  - No tuple formation / deconstruction around pullbacks

Pullbacks with loops still use heap-allocated tuples

### CODE REVIEW

**Comment by asl** [comment]
Tagging @BradLarson

**Comment by asl** [comment]
@swift-ci please check

**Comment by asl** [comment]
@swift-ci please test

File: lib/SILOptimizer/Differentiation/PullbackCloner.cpp [hunk 1]
```diff
@@ -164,7 +164,18 @@ class PullbackCloner::Implementation final
     assert(pbTupleTyple->getNumElements() == values.size() &&
            "The number of pullback tuple fields must equal the number of "
            "pullback struct element values");
```
Comment (dan-zheng) [inline lib/SILOptimizer/Differentiation/PullbackCloner.cpp:166]:
```suggestion
           "pullback tuple element values");
```
(same below)

File: lib/SILOptimizer/Differentiation/PullbackCloner.cpp [hunk 2]
```diff
@@ -164,7 +164,18 @@ class PullbackCloner::Implementation final
...
-    auto res = pullbackTupleElements.insert({origBB, values});
+    auto res = pullbackTupleElements.insert({origBB, { values.begin(), values.end() }});
+    (void)res;
+    assert(res.second && "A pullback struct element already exists!");
```
Comment (dan-zheng) [inline lib/SILOptimizer/Differentiation/PullbackCloner.cpp:169]:
```suggestion
    assert(res.second && "A pullback tuple element already exists!");
```
(same below – there seem to be other "pullback struct" occurrences, if you want to update those too)

File: lib/SILOptimizer/Differentiation/PullbackCloner.cpp [hunk 3]
```diff
@@ -164,7 +164,18 @@ class PullbackCloner::Implementation final
...
+  }
+
+  void initializePullbackTupleElements(SILBasicBlock *origBB,
```
Comment (dan-zheng) [inline lib/SILOptimizer/Differentiation/PullbackCloner.cpp:172]: This method seems to almost entirely duplicate the `SILInstructionResultArray` version of `initializePullbackTupleElements` above.

Would it make sense to rewrite one to call the other? (Perhaps not)

**Review by dan-zheng — COMMENTED** [review]
Looks good to me!

File: lib/SILOptimizer/Differentiation/PullbackCloner.cpp [hunk 4]
```diff
@@ -164,7 +164,18 @@ class PullbackCloner::Implementation final
...
+                                       const llvm::ArrayRef<SILArgument *> &values) {
+    auto *pbTupleTyple = getPullbackInfo().getLinearMapTupleType(origBB);
+    assert(pbTupleTyple->getNumElements() == values.size() &&
+           "The number of pullback tuple fields must equal the number of "
+           "pullback struct element values");
```
Comment (asl) [inline lib/SILOptimizer/Differentiation/PullbackCloner.cpp:177]:
```suggestion
           "pullback tuple element values");
```

File: lib/SILOptimizer/Differentiation/VJPCloner.cpp [hunk 5]
```diff
@@ -223,7 +225,6 @@ class VJPCloner::Implementation final
     auto loc = ri->getOperand().getLoc();
     // Build pullback struct value for original block.
```
Comment (asl) [inline lib/SILOptimizer/Differentiation/VJPCloner.cpp:226]:
```suggestion
    // Build pullback tuple value for original block.
```

File: lib/SILOptimizer/Differentiation/PullbackCloner.cpp (hunk 3 above)
Comment (asl) [reply lib/SILOptimizer/Differentiation/PullbackCloner.cpp:172]: Well, we need two versions here:
 - With `ArrayRef<SILArgument*>` for arguments
 - With `SILInstructionResultArray` (essentially `ArrayRef<SILValue>`) for tuple results

So, the extra conversions here and there very hairy, so I decided to keep the copy, as it is essentially a one-liner :)

**Comment by asl** [comment]
@swift-ci please test

**Review by rxwei — APPROVED** [review]
I skimmed through the changes quickly without looking at the details, but it's a great improvement!

**Comment by asl** [comment]
@rxwei Thanks! Essentially in non-loop case instead of creating / deconstructing the tuples we pass elements directly. The data structures are the same inside – we operate on tuple types.

**Comment by asl** [comment]
@swift-ci please test

**Comment by asl** [comment]
@swift-ci please test

**Comment by asl** [comment]
@swift-ci please test

**Comment by asl** [comment]
@swift-ci please test

**Comment by asl** [comment]
@swift-ci please test

**Comment by asl** [comment]
@swift-ci please test

**Comment by asl** [comment]
@swift-ci please test

**Pull request merged**